### PR TITLE
Scheduler

### PR DESCRIPTION
**Pull request by tristpinsm**
The scheduler appears to be working. Are unit tests necessary for this?

### CODE REVIEW

**Comment by nritsche** [comment]
Yeah, we should really build a unit testing framework. For now, you could add the scheduling functionality to one of the endpoint conf files and run `tests/simulate-chime/run_test.sh` (it should only need a redis client running and automate the rest).

File: coco/master.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def _start_server(self):
         """Start a sanic server."""
         debug = self.log_level == "DEBUG"
         app.run(
-            host="0.0.0.0", port=self.port, workers=self.n_workers, debug=debug, access_log=debug
+            host=self.host, port=self.port, workers=self.n_workers, debug=debug, access_log=debug
```
Comment (nritsche) [inline coco/master.py]: Why?

File: coco/endpoint.py [hunk 2]
```diff
@@ -251,12 +252,12 @@ def client_call(self, host, port, args):
             Is expected to include all values of the endpoint.
         """
         data = copy(self.values)
-        if data:
+        if data and args:
```
Comment (nritsche) [inline coco/endpoint.py]: The scheduler shouldn't be able to call an endpoint that specifies values,right?
In that case `if data:` should have been enough and maybe the scheduler should make sure that none of its endpoints has `.values`

File: coco/master.py [hunk 3]
```diff
@@ -84,6 +94,7 @@ def __init__(self, config_path):
         self.slacker = SlackExporter(self.slack_url)
         config["endpoints"] = self._load_endpoints()
         self._register_config(config)
+        self._start_scheduler()
```
Comment (nritsche) [inline coco/master.py]: Can starting the scheduler here lead to it calling endpoints already? Maybe it should be started last then.

File: coco/master.py [hunk 4]
```diff
@@ -183,6 +196,7 @@ def _load_config(self, config_path):
             self.slack_url = None
             logger.warning("Config variable 'slack_webhook' not found. Slack messaging DISABLED.")
         self.port = config["port"]
+        self.host = config["host"]
```
Comment (nritsche) [inline coco/master.py]: Is this used anywhere?

File: scripts/coco-client [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python3.7
+#!/usr/bin/env python3
```
Comment (nritsche) [inline scripts/coco-client]: Unfortunately this needs to be 3.7 because I used the f-string syntax in a lot of places.
People seemed to be okay with adding a python3.7 installation to where this gets deployed, also we need 3.6+ anyways for sanic.

File: coco/scheduler.py [hunk 6]
```diff
@@ -3,3 +3,139 @@
...
+        """
+        self.host, self.port = host, port
+        # find endpoints to schedule
+        self._gen_timers(endpoints)
+
+    async def start(self):
+        """Start the scheduler (async)."""
+        self.tasks = []
+        for timer in self.timers:
+            logger.info(f"Setting timer '{timer.name}' every {timer.period} s.")
```
Comment (nritsche) [inline coco/scheduler.py]: better debug log level?

File: coco/scheduler.py [hunk 7]
```diff
@@ -3,3 +3,139 @@
...
+class EndpointTimer(Timer):
+    """Timer that calls a coco endopint."""
+
+    def __init__(self, period, endpoint, host, port):
+        self.endpoint = endpoint
+        self.host, self.port = host, port
+        super().__init__(endpoint.name, period)
+
+    async def _call(self):
+        logger.debug(f"{self.name}: {time() - self._start_t}")
```
Comment (nritsche) [inline coco/scheduler.py]: Did you forget this here?

File: coco/scheduler.py [hunk 8]
```diff
@@ -3,3 +3,139 @@
...
+        # Use stored endpoint data
+        data = copy(self.endpoint.values)
```
Comment (nritsche) [inline coco/scheduler.py]: `Endpoint.values` is the configured expected types a request to this endpoint will be checked for before getting forwarded, it does not include the actual data to attach to the request.

As described above, I think the scheduler should (on start) make sure that none of its endpoint has this set, because it wouldn't know what values to use and therefor any call would fail.

File: coco/scheduler.py [hunk 9]
```diff
@@ -3,3 +3,139 @@
...
+        data["coco_report_type"] = self.endpoint.report_type
+
+        # Send request to coco
+        url = f"http://{self.host}:{self.port}/{self.name}"
+        try:
+            async with request(self.endpoint.type, url, data=json.dumps(data)) as r:
+                if r.status != 200:
+                    # TODO send to slack?
+                    logger.error(
+                        f"Scheduled endpoint call ({self.name}) failed: {await r.text()}."
```
Comment (nritsche) [inline coco/scheduler.py]: Both the HTTP error code or the timeout would signal a bug in coco. I think you can just `r.raise_for_status()` and exit coco if `BaseException` is cought.

How coco should represent failures in forwarded requests and what a failure is, is still a bit unclear.

File: coco/master.py (hunk 1 above)
Comment (tristpinsm) [reply coco/master.py]: I saw there was a host option in the config that wasn't being used so I thought it might be an omission. If we don't want this to be configurable, I could instead make it a constant at the top of the file, so that the scheduler can pull it up to know where to send the requests.

File: coco/endpoint.py (hunk 2 above)
Comment (tristpinsm) [reply coco/endpoint.py]: ah, right, I forgot we had discussed that. I'll add that check to the `_gen_timers` method.

File: coco/master.py (hunk 3 above)
Comment (tristpinsm) [reply coco/master.py]: yeah, good point, I will change that.

File: scripts/coco-client (hunk 5 above)
Comment (tristpinsm) [reply scripts/coco-client]: that's fine, what I wanted to change is that it use `/usr/bin/env`. Otherwise it doesn't allow you to use a virtual environment. Would `#!/usr/bin/env python3.7` be ok?

File: coco/master.py (hunk 3 above)
Comment (tristpinsm) [reply coco/master.py]: `_start_server` is blocking, so the best I can do is start the scheduler just before then. It is set up not to call until one period has elapsed, so maybe this won't be a problem.

File: coco/master.py (hunk 1 above)
Comment (nritsche) [reply coco/master.py]: I added the host option for the same reason: to tell the client where the coco server is. According to Richard setting the host parameter here will limit from which interfaces requests are accepted. `0.0.0.0` means no restrictions and that's what we probably want here.

File: coco/master.py (hunk 3 above)
Comment (nritsche) [reply coco/master.py]: Sounds good.

File: scripts/coco-client (hunk 5 above)
Comment (nritsche) [reply scripts/coco-client]: yes

File: coco/master.py (hunk 1 above)
Comment (tristpinsm) [reply coco/master.py]: OK, that makes sense. So I will restore `0.0.0.0` but keep the line that pulls the host from the config to initialise the scheduler. I'll leave a comment here because this had me confused.

Comment (tristpinsm) [reply coco/master.py]: Or I could get rid of the host config parameter in `master` and just use `localhost` to initialise the scheduler.

Comment (nritsche) [reply coco/master.py]: I think the scheduler will by design always run on the same host, so it can use `"localhost"`. Then only coco-client needs to read `host` from the config.

We might want to think about giving the client a separate config.

**Comment by tristpinsm** [comment]
I've addressed @nritsche 's comments, but I haven't tried running it in the `simulate-chime` environment yet.

File: coco/master.py [hunk 10]
```diff
@@ -20,11 +22,18 @@
...
 
 from . import Endpoint, SlackExporter, worker, __version__, RequestForwarder, State
+from .scheduler import Scheduler
 
 app = Sanic(__name__)
 app.config.update({"REDIS": {"address": ("127.0.0.1", 6379)}})
 redis = SanicRedis(app)
-logger = logging.getLogger(__name__)
+
+logger = logging.getLogger()
```
Comment (nritsche) [inline coco/master.py]: What do these changes do?
Why did you remove the `__name__`?
Does the format apply to logging in all coco modules?
I think the log level is set in the config. So shouldn't be hardcoded here.

Comment (tristpinsm) [reply coco/master.py]: I think removing the name ensures that this is getting the root logger. For some reason with `__name__` I wasn't getting any logs from calls in other modules. I thought these should be the same, put apparently not...

The format applies to all logging.

I can get it to set the logging level from the config.

Comment (nritsche) [reply coco/master.py]: I thought that was working already

Comment (tristpinsm) [reply coco/master.py]: with `__name__` I get no logs from the scheduler but without it I do.

Comment (nritsche) [reply coco/master.py]: I guess you have to get the logger after starting the thread

Comment (tristpinsm) [reply coco/master.py]: ok, in the end what was needed was to set the log level in the scheduler. I also moved the formatting and into `__init__.py` so that all modules inherit it.

File: coco/scheduler.py [hunk 11]
```diff
@@ -3,3 +3,129 @@
...
+        self.host, self.port = host, port
+        super().__init__(endpoint.name, period)
+
+    async def _call(self):
+        # logger.debug(f"{self.name}: {time() - self._start_t}")
+
+        # Send request to coco
+        url = f"http://{self.host}:{self.port}/{self.name}"
+        async with request(self.endpoint.type, url) as r:
+            r.raise_for_status()
```
Comment (nritsche) [inline coco/scheduler.py]: Where is this caught?

Comment (tristpinsm) [reply coco/scheduler.py]: It doesn't get caught at the moment. What would you like it to do if this exception is caught?

Comment (nritsche) [reply coco/scheduler.py]: I think it would be best to catch `BaseException`'s from these lines and then write a `logger.error` line like "Scheduler failed calling {endpoint}. Has coco's sanic server crashed?" followed by exiting coco.

Comment (nritsche) [reply coco/scheduler.py]: and `: {e}`

Comment (tristpinsm) [reply coco/scheduler.py]: OK, I have that in now, but since the scheduler runs in a thread it can't exit coco itself. We could have it post a slack warning instead.

Comment (nritsche) [reply coco/scheduler.py]: Yeah, the slack messages we plan to hook up with the logging. So `error` will always go to slack. But also we have to synchronize exiting between all threads and processes. We can leave it like that in this PR, though, I think.
